In [ ]:
! pip install huggingface_hub

In [ ]:
! pip install backoff

In [ ]:
! wget https://github.com/Dao-AILab/flash-attention/releases/download/v2.6.3/flash_attn-2.6.3+cu123torch2.4cxx11abiFALSE-cp311-cp311-linux_x86_64.whl

--2025-03-05 15:35:27--  https://github.com/Dao-AILab/flash-attention/releases/download/v2.6.3/flash_attn-2.6.3+cu123torch2.4cxx11abiFALSE-cp311-cp311-linux_x86_64.whl
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/494232964/a625cc90-5b6f-466c-a0d3-7e5b7acf373b?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250305%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250305T153528Z&X-Amz-Expires=300&X-Amz-Signature=648ee56a53887bddba6312944cdebf1a28da952024df750d87fb2fafaae7fd6f&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dflash_attn-2.6.3%2Bcu123torch2.4cxx11abiFALSE-cp311-cp311-linux_x86_64.whl&response-content-type=application%2Foctet-stream [following]
--2025-03-05 15:35:28--  https://objects.githubusercontent.com/g

In [ ]:
! pip install --no-dependencies --upgrade flash_attn-2.6.3+cu123torch2.4cxx11abiFALSE-cp311-cp311-linux_x86_64.whl

Processing ./flash_attn-2.6.3+cu123torch2.4cxx11abiFALSE-cp311-cp311-linux_x86_64.whl


In [ ]:
! git lfs install
! git clone https://huggingface.co/microsoft/Phi-4-mini-instruct

Git LFS initialized.
Cloning into 'Phi-4-mini-instruct'...
remote: Enumerating objects: 64, done.
remote: Counting objects: 100% (61/61), done.
remote: Compressing objects: 100% (61/61), done.
remote: Total 64 (delta 26), reused 0 (delta 0), pack-reused 3 (from 1)
Unpacking objects: 100% (64/64), 2.68 MiB | 1.86 MiB/s, done.
Filtering content: 100% (3/3), 3.15 GiB | 69.22 MiB/s, done.
Encountered 1 file(s) that may not have been copied correctly on Windows:
	model-00001-of-00002.safetensors

See: `git lfs help smudge` for more details.


In [ ]:
! pip install transformers -U

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.0/44.0 kB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.0/10.0 MB 94.6 MB/s eta 0:00:00
  Attempting uninstall: transformers
    Found existing installation: transformers 4.48.3
    Uninstalling transformers-4.48.3:
      Successfully uninstalled transformers-4.48.3


In [ ]:
AGENT_TOOLS = {
      "booking_fight": {
          "name": "booking_fight",
          "description": "booking fight",
          "parameters": {
              "departure": {
                  "description": "The name of Departure airport code",
                  "type": "str",
              },
              "destination": {
                  "description": "The name of Destination airport code",
                  "type": "str",
              },
              "outbound_date": {
                  "description": "The date of outbound flight",
                  "type": "str",
              },
              "return_date": {
                  "description": "The date of return flight",
                  "type": "str",
              }
          }
      },
      "booking_hotel": {
          "name": "booking_hotel",
          "description": "booking hotel",
          "parameters": {
              "query": {
                  "description": "The name of the city",
                  "type": "str",
              },
              "check_in_date": {
                  "description": "The date of check in",
                  "type": "str",
              },
              "check_out_date": {
                  "description": "The date of check out",
                  "type": "str",
              }
          }
      },
}

In [ ]:
GOOGLE_SEARCH_API_KEY='Your api key from serpapi'

In [ ]:
GOOGLE_SEARCH_ENDPOINT = 'https://serpapi.com/search'

In [ ]:
import requests

In [ ]:
def booking_hotel(query:str,check_in_date: str, check_out_date: str):
        params = {
            "engine": "google_hotels",
            "q": query,
            "check_in_date": check_in_date,
            "check_out_date": check_out_date,
            "adults": "2",
            "currency": "USD",
            "gl": "us",
            "hl": "en",
            "api_key": GOOGLE_SEARCH_API_KEY
        }

        response = requests.get(GOOGLE_SEARCH_ENDPOINT, params=params)
        if response.status_code == 200:
            response = response.json()
            return response["properties"]
        else:
            return None

In [ ]:
def booking_fight(departure: str, destination: str, outbound_date: str, return_date: str):
        go_params = {
            "engine": "google_flights",
            "departure_id": departure,
            "arrival_id": destination,
            "outbound_date": outbound_date,
            "return_date": outbound_date,
            "currency": "USD",
            "hl": "en",
            "api_key": GOOGLE_SEARCH_API_KEY
        }

        # print(params)

        go_response = requests.get(GOOGLE_SEARCH_ENDPOINT, params=go_params)


        result = []

        if go_response.status_code == 200:
            response = go_response.json()
            result.append(response)
            # result += "# outbound \n " + str(response)
        else:
            print('error!!!')
            # return None


        back_params = {
            "engine": "google_flights",
            "departure_id": destination,
            "arrival_id": departure,
            "outbound_date": return_date,
            "return_date": return_date,
            "currency": "USD",
            "hl": "en",
            "api_key": GOOGLE_SEARCH_API_KEY
        }


        back_response = requests.get(GOOGLE_SEARCH_ENDPOINT, params=back_params)



        if back_response.status_code == 200:
            response = back_response.json()
            result.append(response)

        else:
            print('error!!!')
            # return None

        # print(result)

        return result


In [ ]:
SYSTEM_PROMPT = """
    You are my travel agent with some tools available.

"""

In [ ]:
import json

In [ ]:
messages = [
    {
        "role": "system",
        "content": SYSTEM_PROMPT,
        "tools": json.dumps(AGENT_TOOLS), # pass the tools into system message using tools argument
    },
    {
        "role": "user",
        "content": """I have a business trip from London to New York in March 21 2025 to March 27 2025, can you help me to book a hotel and flight tickets"""
    }
]


In [ ]:
import torch
import json
import random
import string
import re
from transformers import AutoModelForCausalLM, AutoProcessor, GenerationConfig,pipeline,AutoTokenizer

In [ ]:
model_path = "Your Phi-4-mini-instruct model path"

In [ ]:
model = AutoModelForCausalLM.from_pretrained(
    model_path,
    device_map="cuda",
    attn_implementation="flash_attention_2",
    torch_dtype="auto",
    trust_remote_code=True)

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

In [ ]:
tokenizer = AutoTokenizer.from_pretrained(model_path, padding_side="left",trust_remote_code=True)

In [ ]:
pipe = pipeline(
    "text-generation",
    model=model,
    tokenizer=tokenizer,
)

Device set to use cuda


In [ ]:
generation_args = {
    "max_new_tokens": 4096,
    "return_full_text": False,
    "temperature": 0.00001,
    "top_p": 1.0,
    "do_sample": True,
}

In [ ]:
messages

[{'role': 'system',
  'content': '\n    You are my travel agent with some tools available.\n\n',
  'tools': '{"booking_fight": {"name": "booking_fight", "description": "booking fight", "parameters": {"departure": {"description": "The name of Departure airport code", "type": "str"}, "destination": {"description": "The name of Destination airport code", "type": "str"}, "outbound_date": {"description": "The date of outbound flight", "type": "str"}, "return_date": {"description": "The date of return flight", "type": "str"}}}, "booking_hotel": {"name": "booking_hotel", "description": "booking hotel", "parameters": {"query": {"description": "The name of the city", "type": "str"}, "check_in_date": {"description": "The date of check in", "type": "str"}, "check_out_date": {"description": "The date of check out", "type": "str"}}}}'},
 {'role': 'user',
  'content': 'I have a business trip from London to New York in March 21 2025 to March 27 2025, can you help me to book a hotel and flight tickets

In [ ]:
output = pipe(messages, **generation_args)

In [ ]:
output

[{'generated_text': '<|tool_call|>[{"name": "booking_fight", "arguments": {"departure": "LHR", "destination": "JFK", "outbound_date": "2025-03-21", "return_date": "2025-03-27"}}, {"name": "booking_hotel", "arguments": {"query": "New York", "check_in_date": "2025-03-21", "check_out_date": "2025-03-27"}}]<|/tool_call|>'}]

In [ ]:
output[0]

{'generated_text': '<|tool_call|>[{"name": "booking_fight", "arguments": {"departure": "LHR", "destination": "JFK", "outbound_date": "2025-03-21", "return_date": "2025-03-27"}}, {"name": "booking_hotel", "arguments": {"query": "New York", "check_in_date": "2025-03-21", "check_out_date": "2025-03-27"}}]<|/tool_call|>'}

In [ ]:
response = output[0]['generated_text']

In [ ]:
response

'<|tool_call|>[{"name": "booking_fight", "arguments": {"departure": "LHR", "destination": "JFK", "outbound_date": "2025-03-21", "return_date": "2025-03-27"}}, {"name": "booking_hotel", "arguments": {"query": "New York", "check_in_date": "2025-03-21", "check_out_date": "2025-03-27"}}]<|/tool_call|>'

In [ ]:
pattern = r'<\|tool_call\|>\[(.*?)\]<\|/tool_call\|>'

match = re.search(pattern, response)

In [ ]:
match

<re.Match object; span=(0, 297), match='<|tool_call|>[{"name": "booking_fight", "argument>

In [ ]:
tool_calls = json.loads(f"[{match.group(1)}]")

In [ ]:
tool_calls

[{'name': 'booking_fight',
  'arguments': {'departure': 'LHR',
   'destination': 'JFK',
   'outbound_date': '2025-03-21',
   'return_date': '2025-03-27'}},
 {'name': 'booking_hotel',
  'arguments': {'query': 'New York',
   'check_in_date': '2025-03-21',
   'check_out_date': '2025-03-27'}}]

In [ ]:
tools_mapping_functions = {
            "booking_fight": booking_fight,
            "booking_hotel": booking_hotel,
}

In [ ]:
result = "{"
for tool_item in tool_calls:
  func_call = tools_mapping_functions[tool_item["name"]]
  tool_reponse = func_call(**tool_item["arguments"])
  if tool_item["name"] == 'booking_fight':
    result += "\"Fights\": \"" + str(tool_reponse) + "\","
  else:
    result += "\"Hotels\": \"" + str(tool_reponse) + "\","

result = result[:-1] + "}"
  # result = result + str(tool_reponse)
  # messages.append({"role": "tool", "content": tool_reponse})

In [ ]:
result

'{"Fights": "[{\'search_metadata\': {\'id\': \'67c86f9f7726dcec26080291\', \'status\': \'Success\', \'json_endpoint\': \'https://serpapi.com/searches/95c8efa8c93dc7a4/67c86f9f7726dcec26080291.json\', \'created_at\': \'2025-03-05 15:37:03 UTC\', \'processed_at\': \'2025-03-05 15:37:03 UTC\', \'google_flights_url\': \'https://www.google.com/travel/flights?hl=en&gl=us&curr=USD&tfs=CBwQAhoeEgoyMDI1LTAzLTIxagcIARIDTEhScgcIARIDSkZLGh4SCjIwMjUtMDMtMjFqBwgBEgNKRktyBwgBEgNMSFJCAQFIAXABmAEB&tfu=EgIIAQ\', \'raw_html_file\': \'https://serpapi.com/searches/95c8efa8c93dc7a4/67c86f9f7726dcec26080291.html\', \'prettify_html_file\': \'https://serpapi.com/searches/95c8efa8c93dc7a4/67c86f9f7726dcec26080291.prettify\', \'total_time_taken\': 2.91}, \'search_parameters\': {\'engine\': \'google_flights\', \'hl\': \'en\', \'gl\': \'us\', \'departure_id\': \'LHR\', \'arrival_id\': \'JFK\', \'outbound_date\': \'2025-03-21\', \'return_date\': \'2025-03-21\', \'currency\': \'USD\'}, \'best_flights\': [{\'flights\

In [ ]:
json.dumps(result)

'"{\\"Fights\\": \\"[{\'search_metadata\': {\'id\': \'67c86f9f7726dcec26080291\', \'status\': \'Success\', \'json_endpoint\': \'https://serpapi.com/searches/95c8efa8c93dc7a4/67c86f9f7726dcec26080291.json\', \'created_at\': \'2025-03-05 15:37:03 UTC\', \'processed_at\': \'2025-03-05 15:37:03 UTC\', \'google_flights_url\': \'https://www.google.com/travel/flights?hl=en&gl=us&curr=USD&tfs=CBwQAhoeEgoyMDI1LTAzLTIxagcIARIDTEhScgcIARIDSkZLGh4SCjIwMjUtMDMtMjFqBwgBEgNKRktyBwgBEgNMSFJCAQFIAXABmAEB&tfu=EgIIAQ\', \'raw_html_file\': \'https://serpapi.com/searches/95c8efa8c93dc7a4/67c86f9f7726dcec26080291.html\', \'prettify_html_file\': \'https://serpapi.com/searches/95c8efa8c93dc7a4/67c86f9f7726dcec26080291.prettify\', \'total_time_taken\': 2.91}, \'search_parameters\': {\'engine\': \'google_flights\', \'hl\': \'en\', \'gl\': \'us\', \'departure_id\': \'LHR\', \'arrival_id\': \'JFK\', \'outbound_date\': \'2025-03-21\', \'return_date\': \'2025-03-21\', \'currency\': \'USD\'}, \'best_flights\': [{\'f

In [ ]:
messages += [
    {
        "role": "assistant",
        "content": output[0]["generated_text"] + ""
    },
    {
    "role": "tool_response",
    "content": json.dumps(result)
   }
]

In [ ]:
messages

[{'role': 'system',
  'content': '\n    You are my travel agent with some tools available.\n\n',
  'tools': '{"booking_fight": {"name": "booking_fight", "description": "booking fight", "parameters": {"departure": {"description": "The name of Departure airport code", "type": "str"}, "destination": {"description": "The name of Destination airport code", "type": "str"}, "outbound_date": {"description": "The date of outbound flight", "type": "str"}, "return_date": {"description": "The date of return flight", "type": "str"}}}, "booking_hotel": {"name": "booking_hotel", "description": "booking hotel", "parameters": {"query": {"description": "The name of the city", "type": "str"}, "check_in_date": {"description": "The date of check in", "type": "str"}, "check_out_date": {"description": "The date of check out", "type": "str"}}}}'},
 {'role': 'user',
  'content': 'I have a business trip from London to New York in March 21 2025 to March 27 2025, can you help me to book a hotel and flight tickets

In [ ]:
output1 = pipe(messages, **generation_args)

In [ ]:
output1

[{'generated_text': "It looks like you've provided a comprehensive list of flight and hotel options for your business trip from London to New York, departing on March 21, 2025, and returning on March 27, 2025. However, the data seems to be a mix of flight and hotel information, and not all of it is directly related to your request.\n\nTo assist you further, I will focus on booking a flight and a hotel for your specified dates. Since I don't have the capability to actually make bookings, I will provide you with a summary of the options based on the data you've given:\n\n**Flight Options:**\n1. A round trip flight from Heathrow (LHR) to John F. Kennedy International Airport (JFK) with various airlines, durations, and carbon emissions. Prices range from $979 to $3350, with different airlines offering different amenities and legroom. Some flights have layovers, while others are direct.\n\n**Hotel Options:**\n1. Pod 51 in New York, a 3-star hotel with a rooftop deck, cafe, bar, and free Wi-

In [ ]:
output1[0]["generated_text"]

"It looks like you've provided a comprehensive list of flight and hotel options for your business trip from London to New York, departing on March 21, 2025, and returning on March 27, 2025. However, the data seems to be a mix of flight and hotel information, and not all of it is directly related to your request.\n\nTo assist you further, I will focus on booking a flight and a hotel for your specified dates. Since I don't have the capability to actually make bookings, I will provide you with a summary of the options based on the data you've given:\n\n**Flight Options:**\n1. A round trip flight from Heathrow (LHR) to John F. Kennedy International Airport (JFK) with various airlines, durations, and carbon emissions. Prices range from $979 to $3350, with different airlines offering different amenities and legroom. Some flights have layovers, while others are direct.\n\n**Hotel Options:**\n1. Pod 51 in New York, a 3-star hotel with a rooftop deck, cafe, bar, and free Wi-Fi, with a total rat

I have found several hotel options for your stay in New York from March 21 to March 27, 2025. Here are some of the best ones:\n\n1. **The Fifty Sonesta Hotel New York**: A 4-star hotel located at 51 St, with an overall rating of 4.1. It offers stylish rooms and suites, a lounge, bar, and free wine hours. The hotel is located near St. Patrick's Cathedral and John F. Kennedy International Airport. The rate per night is $236, with a total rate of $1,417. The hotel is pet-friendly and offers free Wi-Fi, parking, and a fitness center.\n\n2. **The Manhattan at Times Square Hotel**: A 4-star hotel located at 7 Av, with an overall rating of 3.8. It offers modern rooms and suites, a marble-clad lobby lounge, and a gym. The hotel is located near Grand Central Terminal and John F. Kennedy International Airport. The rate per night is $218, with a total rate of $1,307. The hotel is pet-friendly and offers free Wi-Fi, parking, and a restaurant.\n\n3. **The Manhattan by Marriott**: A 4-star hotel located at 7 Av, with an overall rating of 4.1. It offers contemporary rooms and suites, a rooftop lounge, and a fitness center. The hotel is located near Grand Central Terminal and John F. Kennedy International Airport. The rate per night is $238, with a total rate of $1,242. The hotel is pet-friendly and offers free Wi-Fi, parking, and a restaurant.\n\n4. **The Park Hyatt New York**: A 5-star hotel located at 5 Av, with an overall rating of 4.9. It offers luxurious rooms and suites, a rooftop terrace, and a spa. The hotel is located near Grand Central Terminal and John F. Kennedy International Airport. The rate per night is $398, with a total rate of $2,191. The hotel is pet-friendly and offers free Wi-Fi, parking, and a fitness center.\n\n5. **The Langham, New York**: A 5-star hotel located at 7 Av, with an overall rating of 4.9. It offers elegant rooms and suites, a rooftop terrace, and a restaurant. The hotel is located near Grand Central Terminal and John F. Kennedy International Airport. The rate per night is $398, with a total rate of $2,191. The hotel is pet-friendly and offers free Wi-Fi, parking, and a fitness center.\n\nPlease let me know which hotel you would like to book, and I will proceed with the booking process.